FuelLabs/fuel-indexer
# enhancement: add support for unindexable types

### Description

- PR adds support for a `@norelation` directive
  - Applying this directive to any field on an object renders the entire object as `NoRelation`, which means the object is stored as `Json`
    - I chose `Json` over `Blob` because I think it might be easier to extract the item into an object
      - This will definitely require a second pass later
- PR also moves some code around (moves `serialize` and `deserialize` from fuel-indexer-schema to fuel-indexer-lib)
- 

### Testing steps

- [ ] Test was added so CI should pass
- [ ] Build assets `bash scripts/utils/build_test_wasm_module.bash`
- [ ] Start a local fuel node `cargo run -p fuel-node --bin fuel-node`
- [ ] Start the web API `cargo run -p web-api --bin web-api`
- [ ] Start the service with the test indexer 

```bash
cargo run --bin fuel-indexer -- run --manifest packages/fuel-indexer-tests/components/indices/fuel-indexer-test/fuel_indexer_test.yaml --run-migrations
```
- [ ] Trigger some data `curl -X POST http://localhost:8000/block`
- [ ] Open your playground `http://localhost:29987/api/playground/fuel_indexer_test/index1`
- [ ] Make the following query:

```graphql
query {
  usesnotableentity {
    id
    name
    no_table
  }
}
```
- [ ] Should see your type come back as JSON
```json
{
  "data": [
    {
      "id": 1,
      "name": "hello world",
      "no_table": {
        "name": "norelation",
        "size": 1
      }
    }
  ]
}
```
- Note that the item `NoRelationEntity` already comes back as an object straight from the DB, so no further deserialization is needed

### Changelog

- enhancement: add support for a `@norelation` directive that does not create SQL tables for entities
